File: nuc_chain/linkers/lattice_gas.c
#include <stdio.h>
#include <stdlib.h>
#include <math.h>
#include <stdbool.h>

#define NUM_MOVES 4
const double TARGET_ACCEPTANCE = 0.1;
const double KBT = 1;

/* qsort int comparison function */
int int_cmp(const void *a, const void *b)
{
        const int *ia = (const int *)a; // casting pointer types
        const int *ib = (const int *)b;
        return *ia  - *ib;
        /* integer comparison: returns negative if b > a
        *  and positive if a > b */
}

int random_in_range(int n)
{
    long int rv = random();
    double frac = (double)rv/(double)RAND_MAX;
    return (int)(frac*n);
}

void insert_bead(int const *lattice, int *new_lattice, int nbeads, int new_pos)
{
    int i = 0;
    for ( ; i < nbeads && lattice[i] < new_pos; i++) {
        new_lattice[i] = lattice[i];
    }
    new_lattice[i] = new_pos;
    for ( ; i < nbeads; i++) {
        new_lattice[i+1] = lattice[i];
    }
}

int add_move(int const *lattice, int *new_lattice, int nbeads,
        const int max_beads, const int n)
{
    if (nbeads == max_beads) {
        for (int i = 0; i < nbeads; i++) {
            new_lattice[i] = lattice[i];
        }
        return nbeads;
    }
    int new_pos = random_in_range(n);
    insert_bead(lattice, new_lattice, nbeads, new_pos);
    return nbeads + 1;
}

void remove_bead(int const *lattice, int *new_lattice, int nbeads, int del_i)
{
    int i = 0;
    for ( ; i < del_i; i++) {
        new_lattice[i] = lattice[i];
    }
    for (; i < nbeads; i++) {
        new_lattice[i] = lattice[i+1];
    }
}

int remove_move(int const *lattice, int *new_lattice, int nbeads)
{
    if (nbeads == 0) {
        return nbeads;
    }
    int del_i = random_in_range(nbeads);
    remove_bead(lattice, new_lattice, nbeads, del_i);
    return nbeads - 1;

}
int joint_slide_move(int const *lattice, int *new_lattice, int nbeads,
        const int n, double mag)
{
    int left = random_in_range(n);
    int right = random_in_range(n);
    if (right < left) {
        int tmp = left;
        left = right;
        right = tmp;
    }
    int offset = (int) -mag*log(random()/(double)RAND_MAX);
    if (random_in_range(2) == 1) {
        offset *= -1;
    }
    for (int i = 0; i < nbeads; i++) {
        new_lattice[i] = lattice[i];
        if (left <= i && i < right) {
            new_lattice[i] += offset;
            if (new_lattice[i] >= nbeads) {
                new_lattice[i] = nbeads - 1;
            } else if (new_lattice[i] < 0) {
                new_lattice[i] = 0;
            }
        }
    }
    qsort(new_lattice, nbeads, sizeof(int), int_cmp);
    return nbeads;
}

void move_bead(int const *lattice, int *new_lattice, int nbeads, int ind, int new_pos)
{
    // TODO a "jump" based search that uses the offset and assumes packed
    // lattice to make much better initial guess
    int new_ind = ind;
    if (lattice[ind] < new_pos) {
        while (new_ind < nbeads && lattice[new_ind] < new_pos) {
            new_ind++;
        }
        for (int i = 0; i < ind; i++) {
            new_lattice[i] = lattice[i];
        }
        for (int i = ind; i < new_ind - 1; i++) {
            new_lattice[i] = lattice[i+1];
        }
        new_lattice[new_ind-1] = new_pos;
        for (int i = new_ind; i < nbeads; i++) {
            new_lattice[i] = lattice[i];
        }
    } else if (new_pos < lattice[ind]) {
        while (new_ind > 0 && new_pos < lattice[new_ind]) {
            new_ind--;
        }
        for (int i = 0; i < new_ind; i++) {
            new_lattice[i] = lattice[i];
        }
        new_lattice[new_ind] = new_pos;
        for (int i = new_ind + 1; i < ind + 1; i++) {
            new_lattice[i] = lattice[i-1];
        }
        for (int i = ind + 1; i < nbeads; i++) {
            new_lattice[i] = lattice[i];
        }
    }
    /* got too confusing, gonna copy my python code instead
    if (ind == new_ind) {
        for (int i = 0; i < nbeads; i++) {
            new_lattice[i] = lattice[i];
        }
    }
    // now new_ind holds the index we would have to insert before (in [0,
    // new_ind]) to keep array in order
    int i = 0;
    int i_n = 0;
    while (i_n < nbeads) {
        if (i_n == new_ind) {
            new_lattice[i_n++] = new_pos;
            continue;
        }
        if (i == ind) {
            i++;
        }
        new_lattice[i_n++] = lattice[i++];
    }
    */
}

int single_slide_move(int const *lattice, int *new_lattice, int nbeads,
        const int n, double mag)
{
    int choice = random_in_range(nbeads);
    int offset = (int) -mag*log(random()/(double)RAND_MAX);
    if (random_in_range(2) == 1) {
        offset *= -1;
    }
    int new_pos = lattice[choice] + offset;
    if (new_pos < 0) new_pos = 0;
    if (new_pos >= n) new_pos = n-1;
    move_bead(lattice, new_lattice, nbeads, choice, new_pos);
    return nbeads;
}

bool colliding(int const * lattice, int nbeads, int width)
{
    for (int i = 1; i < nbeads; i++) {
        if (lattice[i] - lattice[i-1] < width) {
            return true;
        }
    }
    return false;
}

void print_lattice(int *lattice, int nbeads)
{
    // printf("Lattice<%d>: ", nbeads);
    int i;
    for (i = 0; i < nbeads - 1; i++) {
        printf("%d, ", lattice[i]);
    }
    printf("%d", lattice[i]);
    printf("\n");
}

bool check_lattice_match(int *l1, int *l2, int n)
{
    for (int i = 0; i < n; i++) {
        if (l1[i] != l2[i]) {
            return false;
        }
    }
    return true;
}

void check_test(int *l1, int *l2, int n)
{
    if (!check_lattice_match(l1, l2, n)) {
        print_lattice(l1, n);
        print_lattice(l2, n);
    }
}

void run_tests(void)
{
    int lattice[6] = {2, 5, 17, 21, 33, 0};
    int test_lattice[6];
    int nbeads = 5;

    int expected_lattice_1[6] = {5, 7, 17, 21, 33, 0}; // slide [0]:->7
    move_bead(lattice, test_lattice, nbeads, 0, 7);
    check_test(test_lattice, expected_lattice_1, nbeads);

    int expected_lattice_2[6] = {1, 2, 5, 17, 21, 0}; // slide [4]:->1
    move_bead(lattice, test_lattice, nbeads, 4, 1);
    check_test(test_lattice, expected_lattice_2, nbeads);

    int expected_lattice_3[6] = {2, 17, 21, 33, 42, 0};// slide [1]:->42
    move_bead(lattice, test_lattice, nbeads, 1, 42);
    check_test(test_lattice, expected_lattice_3, nbeads);

    int expected_lattice_4[6] = {1, 2, 5, 17, 21, 33}; // add 1
    insert_bead(lattice, test_lattice, nbeads, 1);
    check_test(test_lattice, expected_lattice_4, nbeads+1);

    int expected_lattice_5[6] = {2, 5, 17, 20, 21, 33};// add 20
    insert_bead(lattice, test_lattice, nbeads, 20);
    check_test(test_lattice, expected_lattice_5, nbeads+1);

    int expected_lattice_6[6] = {2, 5, 17, 21, 33, 35}; // add 35
    insert_bead(lattice, test_lattice, nbeads, 35);
    check_test(test_lattice, expected_lattice_6, nbeads+1);

    int expected_lattice_7[6] = {2, 5, 17, 33, 0, 0}; // remove 3
    remove_bead(lattice, test_lattice, nbeads, 3);
    check_test(test_lattice, expected_lattice_7, nbeads-1);

}

int main(int argc, char *argv[])
{
    /* read inputs */
    if (argc == 2) {
        run_tests();
        return 0;
    }
    if (argc != 6) {
        fprintf(stderr, "usage: %s n_bins particle_width mu num_steps save_interval\n", argv[0]);
        exit(1);
    }
    const int n = strtol(argv[1], NULL, 10);
    const int w = strtol(argv[2], NULL, 10);
    const double mu = strtod(argv[3], NULL);
    const int num_steps = strtol(argv[4], NULL, 10);
    const int save_interval = strtol(argv[5], NULL, 10);
    const int adaptation_interval = (int)fmax(100, fmin(num_steps/100, 10000));
    /* initialize tracking variables */
    int n_accepted[NUM_MOVES] = {0,0,0,0};
    int n_attempted[NUM_MOVES] = {0,0,0,0};
    int j_a = 0; // track adaptation interval
    int j_s = 0; // track save interval
    double params[NUM_MOVES] = {1,1,1,1};
    /* initialize monte carlo state */
    double acc_prob = 0;
    const int max_beads = n/w;
    int lattice[max_beads], new_lattice[max_beads];
    lattice[0] = 0;
    new_lattice[0] = 0;
    for (int k = 1; k < max_beads; k++) {
        lattice[k] = lattice[k-1] + w;
        new_lattice[k] = new_lattice[k-1] + w;
    }
    int nbeads = max_beads;
    int nbeads_new = max_beads;
    srandom((unsigned int)1);
    int rv = 0;
    for (int i = 0; i < num_steps; i++)
    {
        // perform parameter adaptation
        if (j_a == adaptation_interval) {
            printf("Acceptance ratios: ");
            for (int k = 0; k < NUM_MOVES; k++) {
                // make more stringent
                double acceptance_ratio = (double)n_accepted[k]/(double)n_attempted[k];
                printf("%04.8g ", acceptance_ratio);
                if (acceptance_ratio < TARGET_ACCEPTANCE) {
                    params[k]++;
                } else {
                    params[k]--;
                }
                n_accepted[k] = 0;
                n_attempted[k] = 0;
            }
            printf("\n");
            j_a = 0;
        }
        rv = random_in_range(NUM_MOVES);
        switch (rv) {
            case 0:
                nbeads_new = add_move(lattice, new_lattice, nbeads, max_beads, n);
                // new_energy - energy = -1 (one more bead added)
                acc_prob = n/(nbeads+1)*exp(-KBT*(-mu - 1));
                break;
            case 1:
                nbeads_new = remove_move(lattice, new_lattice, nbeads);
                // new_energy - energy = 1 (one bead removed)
                acc_prob = (nbeads/n)*exp(-KBT*(mu + 1));
                break;
            case 2:
                nbeads_new = single_slide_move(lattice, new_lattice, nbeads, n, params[2]);
                acc_prob = 1; // no change in energy
                break;
            case 3:
                nbeads_new = joint_slide_move(lattice, new_lattice, nbeads, n, params[3]);
                acc_prob = 1; // no change in energy
                break;
            default:
                fprintf(stderr, "Invalid move requested.\n");
                exit(1);
        }
        n_attempted[rv]++;
        if (!colliding(new_lattice, nbeads_new, w) && random()/(double)RAND_MAX < acc_prob) {
            n_accepted[rv]++;
            for (int k = 0; k < nbeads_new; k++) {
                lattice[k] = new_lattice[k];
            }
            nbeads = nbeads_new;
        }
        //TODO figure out most satisfyign way to space interval
        // maybe make an option for wehtehr to output initial condition?
        if (j_s == save_interval) {
            print_lattice(lattice, nbeads);
            j_s = 0;
        }
        j_a++; j_s++;
    }

    return 0;
}

File: nuc_chain/linkers/lattice_gas.py
"""Some tools for simulating particle occupancy on a lattice where they can
bind/unbind.

Written to investigate nucleosome positioning data."""

from abc import ABCMeta, abstractmethod
import numpy as np
import pandas as pd

move_history = []

class Lattice(object):
    """A n-bin lattice filled with objects of width w at positions \vec{p}.
    Keeps positions ordered."""
    def __init__(self, n, w, p=None):
        self.n = n
        self.w = w
        self.p = p
        if self.p is None:
            self.p = np.array([])
        self.np = len(self.p)

    def __repr__(self):
        return "Lattice<n={},w={}>: ".format(self.n,self.w) + str(self.p)

    def copy(self):
        return Lattice(self.n, self.w, self.p.copy())

    def add_bead(self, pos):
        """Add as bead at lattice position pos."""
        # assert(1 <= pos <= self.n)
        i = np.searchsorted(self.p, pos)
        self.p = np.insert(self.p, i, pos)
        self.np += 1

    def remove_bead(self, i):
        """Delete the i'th bead."""
        self.p = np.delete(self.p, i)
        self.np -= 1

    def move_bead(self, i, offset):
        """Move the i'th bead by offset. Overshooting drops the bead at the end
        of the lattice.

        Should be equivalent to:
            >>> self.p[i] += offset
            >>> self.p.sort()

        Tests:
            In [27]: l = lg.Lattice(10, 2)
            In [28]: l.p = [6,7,8]
            In [29]: l.move_bead(0, 2); print(l)
            Lattice<n=10,w=2>: [7, 8, 8]
            In [30]: l.move_bead(2, -1); print(l)
            Lattice<n=10,w=2>: [7, 7, 8]
            In [31]: l.move_bead(1, -1); print(l)
            Lattice<n=10,w=2>: [6, 7, 8]
            In [32]: l.move_bead(1, -10); print(l)
            Lattice<n=10,w=2>: [1, 6, 8]
            In [33]: l.move_bead(0, 10); print(l)
            Lattice<n=10,w=2>: [6, 8, 10]
        """
        if offset == 0:
            return
        new_pos = max(1, min(self.n, self.p[i] + offset))
        # get index before which you would insert new_pos if just adding bead
        i_new = np.searchsorted(self.p, new_pos)
        # no movement
        if i <= i_new <= i + 1:
            self.p[i] = new_pos
        # make
        # 1, 2, 3, ..., i-1, i  , i+1, i+2, ..., i+X-2, i+X-1, i+X,...
        # into
        # 1, 2, 3, ..., i-1, i+1, i+2, i+3, ..., i+X-1,     i, i+X,...
        elif i + 1 < i_new:
            X = i_new - i
            self.p[i:i+X-1] = self.p[i+1:i+X]
            # we've slid the elements to the left, so the spot to the left
            # of i+X=i_new opens up
            self.p[i_new-1] = new_pos
        # make
        # 1, 2, 3, ..., i-X-1, i-X, i-X+1, ..., i-1,   i, i+1, ...
        # into
        # 1, 2, 3, ..., i-X-1,   i, i-X,   ..., i-2, i-1, i+1,...
        elif i_new < i:
            X = i - i_new
            self.p[i-X+1:i+1] = self.p[i-X:i]
            # we slide elements to the right, so the spot exact at i-X=i_new
            # opens up
            self.p[i_new] = new_pos

    def move_range(self, left, right, offset):
        """Move the left'th thru right'th beads (inclusize) by offset.
        Overshooting drops the beads at the end of the lattice."""
        if offset == 0:
            return
        self.p[left:right+1] = np.maximum(1, np.minimum(self.n,
                self.p[left:right+1] + offset))
        self.p.sort()
        #TODO generalize move_bead
        # new_pos = np.maximum(1, np.minimum(self.n, self.p[left:right+1] + offset))
        # # get index before which you would insert new_pos if just adding bead
        # left_new, right_new = np.searchsorted(self.p, [new_pos[0], new_pos[-1]])
        # # no movement
        # if i <= i_new <= i + 1:
        #     self.p[i] = new_pos
        # # make
        # # 1, 2, 3, ..., i-1, i  , i+1, i+2, ..., i+X-2, i+X-1, i+X,...
        # # into
        # # 1, 2, 3, ..., i-1, i+1, i+2, i+3, ..., i+X-1,     i, i+X,...
        # elif i + 1 < i_new:
        #     X = i_new - i
        #     self.p[i:i+X-1] = self.p[i+1:i+X]
        #     self.p[i_new] = new_pos
        # # make
        # # 1, 2, 3, ..., i-X-1, i-X, i-X+1, ..., i-1,   i, i+1, ...
        # # into
        # # 1, 2, 3, ..., i-X-1,   i, i-X,   ..., i-2, i-1, i+1,...
        # elif i_new < i:
        #     X = i - i_new
        #     self.p[i-X+1:i+1] = self.p[i-X:i]
        #     self.p[i_new] = new_pos

#############{{{
# moves

class AdaptableMove(metaclass=ABCMeta):
    """Shared structure between adaptable monte carlo moves.

    Implement __call__ to subclass.

    Default values in __init__ are
    ```python
        self.adapt_min = 0
        self.adapt_max = init_lattice.np
        self.adapt_step = 1
        self.param = 1
    ```

    By default, easier_moves() === decrement_param(),
                harder_moves() === increment_param().

    Say your move type needs n random variables per move (n can be zero). Then
    override the list random_generators in class scope to contain functions of
    type size -> random_pool (e.g. lambda size: np.random.random_sample(size)).
    Sequential calls to random_source will then cycle through these random
    numbers.

    By default
    >>>random_generators = [np.random.random_sample]
    which generates a single uniform random source for the move type, one value
    for each move (#TODO up to a fixed max pool size).
    """

    random_generators = {'i': np.random.random_sample}

    def __init__(self, init_lattice, n_moves):
        self.init_lattice_ = init_lattice
        self.n_moves = n_moves
        self.adapt_min = 0
        self.adapt_max = init_lattice.np
        self.adapt_step = 1
        self.param = 1
        self.random_pools_ = {name: rv((n_moves,)) for name, rv in
                              self.__class__.random_generators.items()}
        self.rand_i_ = {name: 0 for name in self.random_pools_}

    def decrement_param(self):
        """Decrement self.param on the grid defined by self.adapt_[min/step/max]."""
        if self.adapt_min < self.param:
            self.param -= self.adapt_step

    def increment_param(self):
        """Increment self.param on the grid defined by self.adapt_[min/step/max]."""
        if self.param < self.adapt_max:
            self.param += self.adapt_step

    def easier_moves(self):
        self.decrement_param()

    def harder_moves(self):
        self.increment_param()

    def random_source(self, name=None):
        if name is None:
            name = list(self.random_pools_.keys())[0]
        if name not in self.random_pools_:
            raise KeyError("AdaptableMove requested random value from non-existent pool")
        rv = self.random_pools_[name][self.rand_i_[name]]
        self.rand_i_[name] += 1
        return rv

    @abstractmethod
    def __call__(self, lattice):
        """Returns a modified version of lattice, with the AdaptableMove
        applied. Does not work in place."""
        return lattice.copy()

class SingleSlideMove(AdaptableMove):
    random_generators = {'choice': lambda size: np.random.random_sample(size),
                         'magnitude': lambda size: (2*np.random.randint(0, 2, size=size) - 1)
                                     *np.random.exponential(size=size)}

    # # override to take reasonably large steps by default
    # def __init__(self, lattice, n_moves):
    #     super().__init__(lattice, n_moves)
    #     self.param = np.round(lattice.np/10).astype(int)

    def __call__(self, lattice):
        new_lattice = lattice.copy()
        if lattice.np == 0:
            move_history.append(['slide', (np.nan, np.nan)])
            return new_lattice
        # first a uniform
        choice = np.floor(lattice.np*self.random_source('choice')).astype(int)
        # then an exponential
        offset = np.round(self.param*self.random_source('magnitude'))
        new_lattice.move_bead(choice, offset)
        move_history.append(['slide', (choice, offset)])
        return new_lattice

class JointSlideMove(AdaptableMove):
    """Only the size of the slide is a parameter, the beads that slide together
    are chosen by picking two random points inside the interval uniformly."""
    random_generators = {'left': lambda size: np.random.random_sample(size),
                         'right': lambda size: np.random.random_sample(size),
                         'magnitude': lambda size: (2*np.random.randint(0, 2, size=size) - 1)
                                     *np.random.exponential(size=size)}

    # # override to take reasonably large steps by default
    # def __init__(self, lattice, n_moves):
    #     super().__init__(lattice, n_moves)
    #     self.param = np.round(lattice.np/10).astype(int)

    def __call__(self, lattice):
        new_lattice = lattice.copy()
        if lattice.np == 0:
            move_history.append(['shift', (np.nan, np.nan, np.nan)])
            return new_lattice
        # left and right boundaries ``uniformly''
        left = np.floor(lattice.np*self.random_source('left')).astype(int)
        right = np.floor(lattice.np*self.random_source('right')).astype(int)
        if right < left:
            left, right = right, left
        # then an exponential
        offset = np.round(self.param*self.random_source('magnitude'))
        new_lattice.move_range(left, right, offset)
        move_history.append(['shift', (left, right, offset)])
        return new_lattice

class AddMoveMuVT(AdaptableMove):
    def __call__(self, lattice):
        new_lattice = lattice.copy()
        i = np.ceil(lattice.n*self.random_source('i')).astype(int)
        new_lattice.add_bead(i)
        move_history.append(['add', i])
        return new_lattice

class RemoveMoveMuVT(AdaptableMove):
    def __call__(self, lattice):
        new_lattice = lattice.copy()
        if new_lattice.np == 0:
            move_history.append(['remove', np.nan])
            return new_lattice
        i = np.floor(lattice.np*self.random_source('i')).astype(int)
        new_lattice.remove_bead(i)
        move_history.append(['remove', i])
        return new_lattice

# end moves
#############}}}

#############{{{
# energy functions
def constant_lattice_energy(lattice):
    return 1


class PreBuiltStericEnergy(object):
    def __init__(self, n, dEb=1):
        self.energies = -dEb*np.ones((n,))

    def __call__(self, lattice):
        if np.any(np.diff(lattice.p) < lattice.w):
            return np.inf
        else:
            return np.sum(self.energies[(lattice.p-1).astype(int)])

class energy_sterics_only(PreBuiltStericEnergy):
    pass
# def steric_lattice_energy(lattice, dEb=1):
#     """Calculate the energy of the current lattice state if we're in the grand
#     canonical ensemble and the lattice is in the presense of a bath with
#     chemical potential mu."""
#     if np.any(np.diff(lattice.p) < lattice.w):
#         return np.inf
#     else:
#         return -lattice.np*dEb

class energy_with_period(PreBuiltStericEnergy):
    def __init__(self, n, period, dEp=1, dEb=1):
        self.period = period
        self.dEp = dEp
        self.dEb = dEb
        self.bp = np.arange(n)
        self.energies = dEp*np.cos(2*np.pi*self.bp/self.period)/2 - dEb

class energy_from_gc(PreBuiltStericEnergy):
    def __init__(self, gc_frac, dEb=1):
        # more gc -> more energetic benefit
        self.energies = -dEb*np.power(gc_frac, 2)


class energy_from_genes(PreBuiltStericEnergy):
    def __init__(self, n, txStart, txDirection, nfr=150, dEb=1):
        self.forward_genes = txStart[txDirection > 0]
        self.reverse_genes = txStart[txDirection < 0]
        self.energies = -dEb*np.ones((n,))
        for gene in self.forward_genes:
            imin = max(0, gene - nfr)
            self.energies[imin:gene] = np.inf
        for gene in self.reverse_genes:
            imax = min(n, gene + nfr)
            self.energies[gene:imax] = np.inf

class energy_from_gene_and_gc(PreBuiltStericEnergy):
    def __init__(self, gc_frac, txStart, txDirection, nfr=150, dEb=1):
        self.n = len(gc_frac)
        gc = energy_from_gc(gc_frac, dEb=dEb)
        self.energies = gc.energies
        genes = energy_from_genes(self.n, txStart, txDirection, nfr=nfr)
        self.energies[np.isinf(genes.energies)] = np.inf

class energy_from_genes_with_period(PreBuiltStericEnergy):
    def __init__(self, n, period, txStart, txDirection, nfr=150, dEp=1, dEb=1):
        self.energies = energy_from_genes(n, txStart, txDirection, nfr=nfr, dEb=dEb).energies
        self.energies += energy_with_period(n, period, dEp=dEp, dEb=0).energies

# end energy functions
#############}}}

#NOTE one approach for canonical ensemble simulation: fix the number bound, only slide moves
def mcmc_simple(n_moves, lattice, energy_f=None, moves=None, kbT=1):
    if moves is None:
        moves = [single_slide_move, add_remove_move]
    move_choice = np.random.randint(len(moves), size=(n_moves,))
    mc_choice = np.random.rand(size=(n_moves,))
    energy = energy_f(lattice)
    min_energy = energy
    for i in range(n_moves):
        new_lattice = moves[move_choice[i]](lattice)
        new_energy = energy_f(new_lattice)
        if new_energy < energy or np.exp(kbT*(energy - new_energy)) < mc_choice[i]:
            energy = new_energy
            lattice = new_lattice
        if new_energy < min_energy:
            min_lattice = new_lattice
    return min_lattice, lattice

def mcmc_full(n_moves, init_lattice, mu, energy_f=None, move_types=None, kbT=1,
              adaptation_interval=None, target_acceptance=None,
              save_interval=None):
    """Perform MHMC with the given initial object, energy function, mc moves,
    and temperature. Allow the monte carlo moves to adapt their magnitudes
    every adaptation_interval steps, so they optimize for a
    target_acceptance.

    In order to simulate the grand canonical ensemble, use the mu argument to
    set the chemical potential of the bath and correctly rescale the acceptance
    probabilities to maintain detailed balance.

    Since the default move_types require mu to be set to work correctly, mu is
    a required parameter. Set to None for other ensembles.
    """
    if energy_f is None:
        energy_f = steric_lattice_energy
    if move_types is None:
        # best moves we have to work with
        move_types = [AddMoveMuVT, RemoveMoveMuVT, SingleSlideMove, JointSlideMove]
    if adaptation_interval is None:
        adaptation_interval = np.max([100, np.min([n_moves/100, 10000])])
    if target_acceptance is None:
        target_acceptance = 0.1
    # tune the adaptability parameters for the given lattice and number of
    # moves if necessary
    move_choice = np.random.randint(len(move_types), size=(n_moves,))
    moves = [MoveType(init_lattice, n_moves=np.sum(move_choice==k))
             for k, MoveType in enumerate(move_types)]
    # initialize mc state variables
    lattice = init_lattice
    energy = energy_f(init_lattice)
    # and variables to save state history
    save_i = []
    lattices = []
    energies = []
    lattices += [lattice]
    energies += [energy]
    # choose the success of all the moves up front for vectorization
    mc_choice = np.random.random_sample(size=(n_moves,))
    # initialize adaptation state variables
    j_a = 0 # number of moves completed since last adaptation
    j_s = 0 # number of moves completed since last save of state
    n_accepted = np.zeros((len(move_types),)) # track acceptance rate
    n_attempted = np.zeros((len(move_types),))
    for i in range(n_moves):
        if j_a == adaptation_interval:
            for k, move in enumerate(moves):
                if n_accepted[k]/n_attempted[k] > target_acceptance:
                    move.harder_moves()
                else:
                    move.easier_moves()
            print('Acceptance ratios: ' + str(n_accepted/n_attempted) + ', Energy: ' + str(energy) + ' [' + str(i) + '/' + str(n_moves) + ']')
            # reset adaptation counter. a move will in fact be completed before
            # we increment at end of loop, so 0, not -1
            j_a = 0
            n_accepted[:] = 0
            n_attempted[:] = 0
        new_lattice = moves[move_choice[i]](lattice)
        new_energy = energy_f(new_lattice)
        Emu = 0
        if mu is not None:
            # this code is specific to the grand canonical ensemble, assuming
            # we choose insert move or delete move randomly then the location,
            # in order to keep detailed balance, we need (in terms of lattice)
            # insertion => prob = n/(np+1)exp{-kbT*(-mu + new_energy - old)}
            # deletion => prob = np/n*exp{-kbT*(mu + new_energy - old)}
            #NOTE: choosing the lattice site first is very inefficient since
            # there are far fewer particles than lattice sites
            if new_lattice.np > lattice.np: # insertion
                Emu = -np.log(lattice.n/(lattice.np+1))/kbT - mu
            elif new_lattice.np < lattice.np: # deletion
                Emu = -np.log(lattice.np/lattice.n)/kbT + mu
        move_history[-1].append(False)
        if new_energy < energy or mc_choice[i] < np.exp(-kbT*(Emu + new_energy - energy)):
            move_history[-1][-1] = True
            energy = new_energy
            lattice = new_lattice
            if save_interval is None:
                save_i.append(i + 1) # so lattices[0] == init_lattice
                energies += [new_energy]
                lattices += [new_lattice]
            n_accepted[move_choice[i]] += 1
        n_attempted[move_choice[i]] += 1
        if j_s == save_interval:
            save_i.append(i + 1) # so lattices[0] == init_lattice
            energies += [energy]
            lattices += [lattice]
            # reset save_interval counter. no moves will be completed before we
            # increment at end of loop, so to compensate, make -1
            j_s = -1
        j_a += 1
        j_s += 1
    return save_i, energies, lattices

def dyad_occupancy_from_lattices(lattices, noise=0):
    occupancy = np.zeros((lattices[0].n,))
    for lattice in lattices:
        # lattice.p in [1, lattice.n]
        err = np.round(noise*np.random.random_sample(size=lattice.np))
        ix = np.array(lattice.p) - 1 + err
        ix[ix >= lattice.n] = lattice.n - 1
        ix[ix < 0] = 0
        occupancy[ix.astype(int)] += 1
    return occupancy

def acf_from_occupancy(occ, max_lag=None):
    if max_lag is None or max_lag > len(occ):
        max_lag = len(occ)
    occdf = pd.Series(occ)
    return np.array([occdf.autocorr(i) for i in range(max_lag)])

def move_history_frame(moves_df=None):
    if moves_df is None:
        moves_df = move_history
    moves_df = pd.DataFrame(moves_df)
    moves_df.columns = ['move_type', 'parameters', 'success']
    return moves_df

def experiment_uniform_chromosome(N=1000000):
    save_i, energies_big, lattices_big = lg.mcmc_full(10000000,
            init_lattice_big, mu, save_interval=100)
    occupancy = dyad_occupancy_from_lattices(lattices)
    plt.figure()
    plt.plot(occupancy)
    plt.xlabel('genomic coordinate (bp)')
    ocupancy = dyad_occupancy_from_lattices(lattices)
    plt.figure()
    plt.plot(occupancy)
    plt.xlabel('genomic coordinate (bp)')
    plt.ylabel('occupancy, MC-time averaged, 10^6 save points')
    plt.title('Occupancy. big, uniform chromosome, nucleosome MC')
    plt.savefig('plots/big_uniform_occupancy.pdf')
    plt.xlim([0, 10000])
    plt.savefig('plots/big_uniform_occupancy_zoom.pdf')

    acf = acf_from_occupancy(occupancy, max_lag=10000)
    plt.figure()
    plt.plot(acf)
    plt.title('ACF(Occupancy). big, uniform chromosome, nucleosome MC')
    plt.ylabel('ACF(Occupancy)')
    plt.xlabel('lag (bp)')
    plt.savefig('plots/big_uniform_acf.pdf')
    plt.ylabel('occupancy, MC-time averaged, 10^6 save points')
    plt.title('Occupancy. big, uniform chromosome, nucleosome MC')
    plt.savefig('plots/big_uniform_occupancy.pdf')
    plt.xlim([0, 10000])
    plt.savefig('plots/big_uniform_occupancy_zoom.pdf')

    acf = acf_from_occupancy(occupancy, max_lag=10000)
    plt.figure()
    plt.plot(acf)
    plt.title('ACF(Occupancy). big, uniform chromosome, nucleosome MC')
    plt.ylabel('ACF(Occupancy)')
    plt.xlabel('lag (bp)')
    plt.savefig('plots/big_uniform_acf.pdf')





